include/SampleBuffer.h

```diff
@@ -39,320 +48,63 @@
-		}
-
-		bool isBackwards() const
-		{
-			return m_isBackwards;
-		}
-
-		void setBackwards(bool backwards)
-		{
-			m_isBackwards = backwards;
-		}
-
-		int interpolationMode() const
-		{
-			return m_interpolationMode;
-		}
-
-
-	private:
-		f_cnt_t m_frameIndex;
-		const bool m_varyingPitch;
-		bool m_isBackwards;
-		SRC_STATE * m_resamplingData;
-		int m_interpolationMode;
-
-		friend class SampleBuffer;
-
-	} ;
-
-
-	SampleBuffer();
-	// constructor which either loads sample _audio_file or decodes
-	// base64-data out of string
-	SampleBuffer(const QString & audioFile, bool isBase64Data = false);
-	SampleBuffer(const sampleFrame * data, const f_cnt_t frames);
-	explicit SampleBuffer(const f_cnt_t frames);
-	SampleBuffer(const SampleBuffer & orig);
-
-	friend void swap(SampleBuffer & first, SampleBuffer & second) noexcept;
-	SampleBuffer& operator= (const SampleBuffer that);
-
-	~SampleBuffer() override;
-
-	bool play(
-		sampleFrame * ab,
-		handleState * state,
-		const fpp_t frames,
-		const float freq,
-		const LoopMode loopMode = LoopOff
-	);
-
-	void visualize(
-		QPainter & p,
-		const QRect & dr,
-		const QRect & clip,
-		f_cnt_t fromFrame = 0,
-		f_cnt_t toFrame = 0
-	);
-	inline void visualize(
-		QPainter & p,
-		const QRect & dr,
-		f_cnt_t fromFrame = 0,
-		f_cnt_t toFrame = 0
-	)
-	{
-		visualize(p, dr, dr, fromFrame, toFrame);
-	}
-
-	inline const QString & audioFile() const
-	{
-		return m_audioFile;
-	}
-
-	inline f_cnt_t startFrame() const
-	{
-		return m_startFrame;
-	}
-
-	inline f_cnt_t endFrame() const
-	{
-		return m_endFrame;
-	}
-
-	inline f_cnt_t loopStartFrame() const
-	{
-		return m_loopStartFrame;
-	}
-
-	inline f_cnt_t loopEndFrame() const
-	{
-		return m_loopEndFrame;
-	}
-
-	void setLoopStartFrame(f_cnt_t start)
-	{
-		m_loopStartFrame = start;
-	}
-
-	void setLoopEndFrame(f_cnt_t end)
-	{
-		m_loopEndFrame = end;
-	}
-
-	void setAllPointFrames(
-		f_cnt_t start,
-		f_cnt_t end,
-		f_cnt_t loopStart,
-		f_cnt_t loopEnd
-	)
-	{
-		m_startFrame = start;
-		m_endFrame = end;
-		m_loopStartFrame = loopStart;
-		m_loopEndFrame = loopEnd;
-	}
-
-	inline f_cnt_t frames() const
-	{
-		return m_frames;
-	}
+	using value_type = sampleFrame;
+	using reference = sampleFrame&;
+	using const_iterator = std::vector<sampleFrame>::const_iterator;
+	using const_reverse_iterator = std::vector<sampleFrame>::const_reverse_iterator;
+	using difference_type = std::vector<sampleFrame>::difference_type;
+	using size_type = std::vector<sampleFrame>::size_type;
+
+	SampleBuffer() = default;
+	SampleBuffer(const SampleBuffer& other) noexcept;
+	SampleBuffer(SampleBuffer&& other) noexcept;
+	SampleBuffer(const QString& audioFile);
+	SampleBuffer(const QByteArray& base64);
+	SampleBuffer(const sampleFrame* data, int numFrames, int sampleRate = Engine::audioEngine()->processingSampleRate());
+	explicit SampleBuffer(int numFrames);
+	~SampleBuffer() noexcept;
+
+	SampleBuffer& operator=(SampleBuffer other) noexcept;
```

Comment:
You should also add a move assignment operator to meet the rule of five.

Reply:
I was under the assumption that the copy-and-swap idiom handles that for us.

Reply:
The copy-and-swap idiom is used to implement the copy assignment operator, which you did, and it looks correct to me. The move assignment operator is the only one that hasn't been implemented yet.

Reply:
From https://en.cppreference.com/w/cpp/language/operators#Assignment_operator:
``
...the copy-and-swap assignment operator, which takes its parameter by value (thus working as both copy- and move-assignment depending on the value category of the argument), swaps with the parameter, and lets the destructor clean it up. 
``

https://en.cppreference.com/w/cpp/language/rule_of_three also suggests to replace both operators if the copy-and-swap idiom is being used (since the idiom covers both situations).

So from what I understand, this is saying that the operator works both as a copy and move assignment operator.
I might go with using the rule of zero if I can, but I still have questions relating to synchronization (calling a move constructor while the moved object is being played by the ``AudioEngine`` through the play handles could create problems, for example).

Reply:
It looks like you're right.. My bad.

Reply:
> It looks like you're right.. My bad.

No worries :smile:.